### PR TITLE
Apply LXD Project before validating profiles

### PR DESCRIPTION
**Pull request by masnax**
Currently, Juju's handling of LXD projects occurs after validating the default profile [here](https://github.com/juju/juju/blob/e2c7b4c88e516976666e3d0c9479d0d3c704e643/provider/lxd/server.go#L341). 

This is problematic on two fronts. First is that the default profile in the selected project is not being validated. The second issue is for ongoing development of LXD Cloud, which will be unable to use the default project at all, causing an immediate error.

As such, this properly applies the project before making any API request. 

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- ~[ ] [Integration tests](https://github.com/juju/juju/tree/develop/tests), with comments saying what you're testing~
- ~[ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps
`juju add-model <model> <cloud> --config project=<my-lxd-project>`

### CODE REVIEW

File: provider/lxd/environ.go [hunk 1]
```diff
@@ -156,15 +156,11 @@ func (env *environ) SetCloudSpec(_ stdcontext.Context, spec environscloudspec.Cl
 	defer env.lock.Unlock()
 
 	serverFactory := env.provider.serverFactory
-	server, err := serverFactory.RemoteServer(spec)
+	server, err := serverFactory.RemoteServer(spec, env.ecfgUnlocked.project())
```
Comment (SimonRichardson) [inline provider/lxd/environ.go:159]: I'm not a fan of this, I think at the very least we should wrap the spec to become an `LXDCloudSpec` with a `Project` field.

**Review by SimonRichardson — APPROVED** [review]
Thank-you @masnax for this set of changes and for the quick turn around.

Approved.

**Comment by SimonRichardson** [comment]
@masnax there are still some tests failing...

FAIL: upgrader_test.go:301: modelUpgradeSuite.TestUpgradeModelForControllerModelJuju3
FAIL: upgrader_test.go:305: modelUpgradeSuite.TestUpgradeModelForControllerModelJuju3DryRun
FAIL: upgrader_test.go:309: modelUpgradeSuite.TestUpgradeModelForControllerModelJuju3Failed
FAIL: upgrader_test.go:511: modelUpgradeSuite.TestUpgradeModelJuju3
FAIL: upgrader_test.go:515: modelUpgradeSuite.TestUpgradeModelJuju3DryRun
FAIL: upgrader_test.go:519: modelUpgradeSuite.TestUpgradeModelJuju3Failed
FAIL: precheck_test.go:89: SourcePrecheckSuite.TestTargetController3Failed
FAIL: migrate_test.go:69: migrateSuite.TestValidatorsForModelMigrationSourceJuju3
FAIL: migrate_test.go:83: migrateSuite.TestValidatorsForModelMigrationSourceJuju31
FAIL: validation_test.go:383: upgradeValidationSuite.TestGetCheckForLXDVersionFailed
FAIL: validation_test.go:358: upgradeValidationSuite.TestGetCheckForLXDVersionLXD
FAIL: validation_test.go:362: upgradeValidationSuite.TestGetCheckForLXDVersionLocalhost
FAIL: upgrade_test.go:23: upgradeValidationSuite.TestValidatorsForControllerUpgradeJuju3
FAIL: upgrade_test.go:111: upgradeValidationSuite.TestValidatorsForModelUpgradeJuju3

**Comment by masnax** [comment]
> @masnax there are still some tests failing...
> 
> FAIL: upgrader_test.go:301: modelUpgradeSuite.TestUpgradeModelForControllerModelJuju3 FAIL: upgrader_test.go:305: modelUpgradeSuite.TestUpgradeModelForControllerModelJuju3DryRun FAIL: upgrader_test.go:309: modelUpgradeSuite.TestUpgradeModelForControllerModelJuju3Failed FAIL: upgrader_test.go:511: modelUpgradeSuite.TestUpgradeModelJuju3 FAIL: upgrader_test.go:515: modelUpgradeSuite.TestUpgradeModelJuju3DryRun FAIL: upgrader_test.go:519: modelUpgradeSuite.TestUpgradeModelJuju3Failed FAIL: precheck_test.go:89: SourcePrecheckSuite.TestTargetController3Failed FAIL: migrate_test.go:69: migrateSuite.TestValidatorsForModelMigrationSourceJuju3 FAIL: migrate_test.go:83: migrateSuite.TestValidatorsForModelMigrationSourceJuju31 FAIL: validation_test.go:383: upgradeValidationSuite.TestGetCheckForLXDVersionFailed FAIL: validation_test.go:358: upgradeValidationSuite.TestGetCheckForLXDVersionLXD FAIL: validation_test.go:362: upgradeValidationSuite.TestGetCheckForLXDVersionLocalhost FAIL: upgrade_test.go:23: upgradeValidationSuite.TestValidatorsForControllerUpgradeJuju3 FAIL: upgrade_test.go:111: upgradeValidationSuite.TestValidatorsForModelUpgradeJuju3

Ah, sorry about that. I had set up `juju-db` incorrectly and missed these. Should be fixed now.

**Comment by masnax** [comment]
> Thanks for the turn around @masnax, just one last one:
> 
> ```
> FAIL: upgrade_test.go:111: upgradeValidationSuite.TestValidatorsForModelUpgradeJuju3
> ```


Thanks for your patience, hopefully that does it :)

**Comment by masnax** [comment]
@SimonRichardson Just checking in, is there anything else holding this PR back from merging?

**Pull request merged**